op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+```go
+type Deriver interface {
+    OnEvent(ctx context.Context, event *Event)
+}
+```
+A deriver is assumed to read-access to any state it needs. Composed by the constructor of the deriver.
+
+#### Composable derivers
+
+Writes idiomatically happen by implementing a purpose-specific Deriver:
+```go
+type Engine struct {
+  control EngineController
+}
+
+func (e *Engine) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+        case *NewPayload:
+            select {
+              // events can include ways to communicate the result, if necessary, across threads
+              case x.Result <- e.onNewPayload(x.Envelope):
+              case <-ctx.Done():
+            }
+        // *insert other engine events*
+    }
+}
+
+func (e *Engine) onNewPayload(env *eth.ExecutionPayloadEnvelope) error {
+    // ... check prestate first etc., insofar the EngineController does not already maintain it.
+    return e.control.InsertUnsafePayload(ctx, x.PayloadEnvelope)
+}
+```
+
+Other derivers can then re-use this, by composing:
+```go
+type Rollup struct {
+    engine Deriver
+    pipeline Deriver
+    // ...
+}
+
+func (r *Rollup) OnEvent(ctx context.Context, event *Event) {
+  switch x := event.Data.(type) {
+  case *NewPayload:
+      // 1) update some rollup state: ...
+      // 2) and propagate to the engine deriver
+      r.engine.OnEvent(ctx, event)
+  case *L1Change:
+      r.pipeline.OnEvent(ctx, event)
+  default:
+      r.pipeline.OnEvent(ctx, event) // can forward anything to a fallback
+  }
+}
+```
+
+Or, a deriver can be given access to the root-level Deriver, to surface new event data to:
+```go
+type Pipeline struct {
+    root Deriver
+}
+
+func (p *Pipeline) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+    case *L1Change:
+        // generate block attributes and give them to the unsafe-block processor to extend the chain
+        // (or forward to next deriver if already known attributes)
+        p.root.OnEvent(ctx, &Event{Resource: UnsafeL2, Data: AttributesEvent{attributes}})
+    }
+}
+```
+
+#### Deriver design principles
+
+Custom stacks like various Alt-DA implementations can wrap around any existing derivers,
+extend with new events, possibly re-using standard OP-Plasma events, for additional de-duplication of alt-DA code.
+
+To encourage composability, and reduce code, events should be documented and small in scope.
+I.e. create an event for a deriver to turn attributes in an unsafe block,
+and follow that up with events to try to promote the block safety, suggest it for finality, and persisting it to the safe head DB.
+rather than tightly coupling the four acts: (1) block processing, (2) updating of the safe-head, (3) remembering the finality data all at once, (4) persisting the safe head change.
+
+
+#### Threads and parallel derivers
+
+Stacks can define a single `OnEvent` handler, and by using `Resource`, still execute on parallel threads, if desired.
```

Comment:
I'm unclear on how the `OnEvent` implementation is controlled.  In the examples above it looks like events can be targeted to specific derivers (`root` vs `engine` vs `pipeline`) but this suggests that they always call the "stacks" `OnEvent` and it routes based on the `Resource`.

How the stack is created is a really key part of this because fault proofs vs op-node is going to want to control how messages are passed to switch between sync and async, but different stacks will want to plug in different derivers. It would be good to avoid having to have a interop+derive, interop+fp, altda+derive, altda+fp kind of stacks.  So I think we want to decouple how message passing happens from where the message is handled (which this part implies but doesn't seem to fit with some of the stuff above).

Reply:
Good points, will combine into the other resource/event related discussion.